### PR TITLE
Make sure KernelTemplateMapper extractors's size is the same as the number of args

### LINKED ISSUE #531: @ti.kernel couldn't handle a mixed args of primitives and `ti.ext_arr()`
**Describe the bug**

The following code snippet

```py
@ti.kernel
def foo(x: ti.i32, arr: ti.ext_arr()):
  pass

foo(42, np.array([42]))
```

would crash Taichi at `materialize()`. Stack trace:

```
Traceback (most recent call last):
  File "test_empty.py", line 22, in <module>
    test_empty2()
  File "test_empty.py", line 20, in test_empty2
    func(42, arr)
  File "Github/taichi/python/taichi/lang/kernel.py", line 350, in __call__
    self.materialize(key=key, args=args, arg_features=arg_features)
  File "Github/taichi/python/taichi/lang/kernel.py", line 208, in materialize
    visitor.visit(tree)
  File "miniconda3/envs/taichi-dev/lib/python3.7/ast.py", line 271, in visit
    return visitor(node)
  File "Github/taichi/python/taichi/lang/transformer.py", line 449, in visit_Module
    self.generic_visit(node)
  File "Github/taichi/python/taichi/lang/transformer.py", line 78, in generic_visit
    value = self.visit(value)
  File "miniconda3/envs/taichi-dev/lib/python3.7/ast.py", line 271, in visit
    return visitor(node)
  File "Github/taichi/python/taichi/lang/transformer.py", line 484, in visit_FunctionDef
    array_dt = self.arg_features[i][0]
IndexError: tuple index out of range
```

**Reason**

Here's how the arg extractors are created:

https://github.com/taichi-dev/taichi/blob/9c96fb814d3bfb5950411250bed8df3eaeebed9e/python/taichi/lang/kernel.py#L63

It only creates an extractor for args with `extract` attribute, meaning the size of `self.extractors` doesn't always match the number of arguments.

Tracing how those extractors are used, I saw they are used to create `arg_features`, which is passed into `materialize()`, which then crashes at here:

https://github.com/taichi-dev/taichi/blob/9c96fb814d3bfb5950411250bed8df3eaeebed9e/python/taichi/lang/transformer.py#L484-L485

**Possible fix**
The absolutely simple fix is to always make `self.extractors` match the number of the kernel arguments. I had such a fix locally:

```py
class KernelTemplateMapper:

  def __init__(self, annotations, template_slot_locations):
    self.annotations = annotations
    dummy_extract = lambda arg: (type(arg).__name__, )
    self.extractors = tuple((i, anno.extract if hasattr(
        anno, 'extract') else dummy_extract) for (i, anno) in enumerate(self.annotations))
    ...
```

But do you think the fix is appropriate? @yuanming-hu

### PR DESCRIPTION
**Pull request by k-ye**
Fixes #531

### CODE REVIEW

File: tests/python/test_empty.py [hunk 1]
```diff
@@ -8,3 +8,12 @@ def func():
...
   func()
+
+@ti.all_archs
+def test_empty_args():
+
+  @ti.kernel
+  def func(x: ti.i32, arr: ti.ext_arr()):
+    pass
+
+  foo(42, np.arange(10, dtype=np.float32))
```
Comment (digitalillusions) [inline tests/python/test_empty.py]: Did you mean to call func in the test?

Comment (archibate) [reply tests/python/test_empty.py]: Yeah, I think so. `foo` maybe directly copied from the example code of #531.

Comment (k-ye) [reply tests/python/test_empty.py]: Nice catch! Done

**Comment by yuanming-hu** [comment]
Cool, thanks!

**Pull request merged**